File: py/year2022/day03.py
import functools

with open(f"input/year2022/day03.txt", "r") as f:
    lines = f.read().splitlines()

def get_priority(x: str): return ord(x.lower()) - ord("a") + 1 + (26 if x.isupper() else 0)

def get_priority_line(line):
    compartment1, compartment2 = set(line[:len(line) // 2]), set(line[len(line) // 2:])
    both_item = list(compartment1 & compartment2)[0]
    return get_priority(both_item)

def get_priority_group(group):
    all_item = list(functools.reduce(lambda curr, new : curr & set(new), group, set(group[0])))[0]
    return get_priority(all_item)

print(sum([get_priority_line(line) for line in lines]))
print(sum([get_priority_group(group) for group in zip(lines[::3], lines[1::3], lines[2::3])]))

File: rust/src/bin/year2022/day03.rs
use std::collections::HashSet;
use std::fs;

pub fn run() {
    let mut text = fs::read_to_string("../input/year2022/day03.txt").unwrap();
    text = text.replace("\r\n", "\n");
    let lines = text
        .split("\n")
        .filter(|line| line.len() > 0)
        .collect::<Vec<&str>>();

    let part1: i32 = lines
        .clone()
        .into_iter()
        .map(|line| get_priority_line(line))
        .sum();

    let part2: i32 = (0..lines.len())
        .step_by(3)
        .map(|i| get_priority_group(vec![lines[i], lines[i + 1], lines[i + 2]]))
        .sum();

    println!("{}", part1);
    println!("{}", part2);
}

fn get_priority_line(line: &str) -> i32 {
    let divider = line.len() / 2;
    let compartment1 = &line[0..divider];
    let compartment2 = &line[divider..line.len()];

    get_priority_group(vec![compartment1, compartment2])
}

fn get_priority_group(group: Vec<&str>) -> i32 {
    let shared_items = group
        .into_iter()
        .map(|compartment| HashSet::from_iter(compartment.chars()))
        .reduce(|curr: HashSet<char>, new: HashSet<char>| {
            curr.intersection(&new).cloned().collect::<HashSet<char>>()
        })
        .unwrap()
        .into_iter()
        .collect::<String>();

    get_priority(shared_items.as_str())
}

fn get_priority(x: &str) -> i32 {
    let value = if x.to_ascii_uppercase() == x.to_string() {
        26
    } else {
        0
    };
    ord(x.to_ascii_lowercase().as_str()) - ord("a") + 1 + value
}

fn ord(x: &str) -> i32 {
    x.chars().nth(0).unwrap() as i32
}
